services/sts/src/main/java/software/amazon/awssdk/services/sts/internal/StsWebIdentityTokenFileCredentialsProvider.java

```diff
@@ -0,0 +1,192 @@
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.internal.WebIdentityTokenCredentialProperties;
+import software.amazon.awssdk.core.SdkSystemSetting;
+import software.amazon.awssdk.services.sts.StsClient;
+import software.amazon.awssdk.utils.ToString;
+
+
+/**
+ * A credential provider that will read web identity token file path, aws role arn, aws session name
+ * from system properties or environment variables  and StsClient for using
+ * web identity token credentials with STS. If StsClient is not passed or null then it will use Default Sts client.
+ */
+@SdkPublicApi
+public class StsWebIdentityTokenFileCredentialsProvider implements AwsCredentialsProvider {
+
+    private final AwsCredentialsProvider credentialsProvider;
+    private final RuntimeException loadException;
+
+    private StsWebIdentityTokenFileCredentialsProvider(BuilderImpl builder) {
+        AwsCredentialsProvider credentialsProviderLocal = null;
+        RuntimeException loadExceptionLocal = null;
+
+        try {
+            Path webIdentityTokenFile =
+                builder.webIdentityTokenFile != null ? builder.webIdentityTokenFile
+                                                     : Paths.get(trim(SdkSystemSetting.AWS_WEB_IDENTITY_TOKEN_FILE
+                                                                          .getStringValueOrThrow()));
+
+            String roleArn = builder.roleArn != null ? builder.roleArn
+                                                     : trim(SdkSystemSetting.AWS_ROLE_ARN.getStringValueOrThrow());
+
+            Optional<String> roleSessionName =
+                builder.roleSessionName != null ? Optional.of(builder.roleSessionName)
+                                                : SdkSystemSetting.AWS_ROLE_SESSION_NAME.getStringValue();
+
+            WebIdentityTokenCredentialProperties credentialProperties =
+                WebIdentityTokenCredentialProperties.builder()
+                                                    .roleArn(roleArn)
+                                                    .roleSessionName(roleSessionName.orElse(null))
+                                                    .webIdentityTokenFile(webIdentityTokenFile)
+                                                    .build();
+
+            credentialsProviderLocal = new StsWebIdentityCredentialsProviderFactory().create(credentialProperties,
+                    builder.stsClient);
+        } catch (RuntimeException e) {
+            // If we couldn't load the credentials provider for some reason, save an exception describing why. This exception
+            // will only be raised on calls to getCredentials. We don't want to raise an exception here because it may be
+            // expected (eg. in the default credential chain).
+            loadExceptionLocal = e;
+        }
```

Comment:
Do we need to use the StsWebIdentityCredentialsProviderFactory here? Can we just move the logic from StsWebIdentityCredentialsProviderFactory into this class and call this class from that one? It's hard for me to wrap my head around these factory-generates-a-delegate-for-me-to-use approach.

Reply:
Agree, moved it now